### ISSUE TITLE
[FEATURE] - Reuse embeddings

### ISSUE THREAD

**Issue by edeandrea** [langchain4j/langchain4j#1547]
**Is your feature request related to a problem? Please describe.**
It would be really nice to be able to re-use embeddings across application instance restarts, especially in the inner loop cycle. For example, if my application computes embeddings on startup I may only want to do that once while I am working on a feature. Sometimes the computation can be time consuming. The only current solution is to switch to an `InMemoryEmbeddingStore`, which already supports this feature.

But what if I want this feature when using some other embedding store? Currently its not possible. And maybe it won't be possible for all store implementations, in which case thats fine. But each store should be allowed the option to provide that functionality.

**Describe the solution you'd like**

Currently the `InMemoryEmbeddingStore` implements some methods that solve this problem.  There are several `serializeTo...` methods that write out the contents of the embedding store.

I'd like to move these methods "up" to the `EmbeddingStore` interface. I would make default implementations of them which throw some unsupported exception, so it would preserve backwards-compatibility. With this particular change, only the `InMemoryEmbeddingStore` would continue to implement the functionality.

But future enhancements/PRs could implement the functionality for other embedding stores.

One challenge I see though is that `from...` methods which populate an embedding store from some other source are declared `static` in `InMemoryEmbeddingStore`. We'd have to potentially deprecate those and introduce instance-level ones, or introduce some kind of factory that knew how to instantiate the correct implementation type.

Again, though, everything should be done in a backwards-compatible manner and no other embedding store providers would be touched as part of this change.

**Additional context**

If we all agree I'm happy to take this on and work on it. Just wanted to propose it before spending time on it to gather feedback.

**Comment by LizeRaes** [comment]
Thank you @edeandrea for your suggestion! ```Embeddings``` are indeed calculation-heavy and typically calculated only once in the application lifecycle, and then updated when the source documents change. 
Do I understand correctly that you would like to have a way to persist ```EmbeddingStores``` over instance restarts? I believe all ```EmbeddingStores``` except the ```InMemoryEmbeddingStore``` are persisted automatically and behave mostly like a normal database.
The main advantage I see in adding ```serializeTo...``` and ```from...``` methods also for the other types of ```EmbeddingStores```, is to have a common intermediary format to mirgate ```Embeddings``` from one ```EmbeddingStore``` implementation to another. Is that what you are after?

**Comment by edeandrea** [comment]
Thanks for the note @LizeRaes !

That's not exactly what I was looking for but is a nice thought! 

I agree, that in production this feature would most likely not be used, as a "real" embedding store would maintain the embeddings between application restarts. 

The use case I'm solving for is the inner loop development cycle. Perhaps the user uses testcontainers/docker compose/etc to stand up a "real" embedding store while they are doing dev. Each time testcontainers/docker compose shuts down, the computed embeddings are gone. This feature would allow the embeddings to be written to the filesystem so that they don't need to be re-computed. My initial thought was that the format of what was written would be specific to the embedding store vendor, but it doesn't have to be. To be honest I haven't looked into depth on that. 

I am being somewhat selfish too, because my main purpose for this feature is with the quarkus LangChain4j extension, which would use this feature in its dev mode/live coding. 

Currently if I'm using a real embedding store, each time I make a code change it needs to recompute the embeddings and re-store them. 

I've already implemented this feature for the in-memory embedding store, because that store already exposed the capability. I'd like to be able to extend that to other embedding stores, although it would be in separate PRs should someone decide to actually implement the capabilities for other embedding store implementations. I wouldn't be tackling that in this change. I just want to have a consistent interface for now.

See https://docs.quarkiverse.io/quarkus-langchain4j/dev/easy-rag.html#_reusing_embeddings

Does that make sense? Again, probably not a feature that will be widely used in production, but definitely a time-saver in inner-loop development.

**Comment by InAnYan** [comment]
> each time I make a code change it needs to recompute the embeddings and re-store them

This sounds strange, why? There are only 2 reasons where you need to recompute embeddings:
- You chose a different embedding model (embeddings generated from different models are incompatible).
- You change the chunking algorithm.

>  Each time testcontainers/docker compose shuts down, the computed embeddings are gone.

Sorry, I'm not an expert in Docker, but I see N ways to solve this:
- You should use a real vector database.
- Then, you look up how the data is persisted on disk.
- And when Docker container is shutdown, before that do a backup.

For example, there are `pgvector` extension for `PostgreSQL` and `sqlite-vss` extension for `SQLite`. And you can very easy backup the results in an `.sql` file dump.

You can also not use a Docker container, but start it locally on your computer.

However, the easiest method is to find an **embedded vector database** (just like SQLite is an embedded database). That would be like a library and store everything in files. But unfortunately, every vector database requires a dedicated server

**Comment by edeandrea** [comment]
> > each time I make a code change it needs to recompute the embeddings and re-store them
> 
> This sounds strange, why? There are only 2 reasons where you need to recompute embeddings:
> 
> * You chose a different embedding model (embeddings generated from different models are incompatible).
> * You change the chunking algorithm.

I agree 100% that _in production_ this is how things will work. As I stated, this feature is more meant for local development/inner loop.

> 
> > Each time testcontainers/docker compose shuts down, the computed embeddings are gone.
> 
> Sorry, I'm not an expert in Docker, but I see N ways to solve this:
> 
> * You should use a real vector database.
> * Then, you look up how the data is persisted on disk.
> * And when Docker container is shutdown, before that do a backup.

This is just not how the real world works. When a docker container shuts down, data is gone, especially when using ephemeral filesystems. Application frameworks like Testcontainers make using "real" things (vector databases, relational databases, message brokers, etc) trivial during development time. The part that makes it trivial is that the containers are ephemeral.

Think about a relational database for example (Postgres/MySQL/Oracle/etc). During development, if I start up a database container, my application will need to create a scheme and potentially pre-load some data. Frameworks like Hibernate support this - the data I want to pre-load is available on the filesystem and things just work. _In production_, the "real" database will already be running somewhere, the schema would have already been created, and data would already exist. What I'm trying to do here is the same thing, conceptually. I'm trying to solve a time-consuming use case when doing development.

> 
> For example, there are `pgvector` extension for `PostgreSQL` and `sqlite-vss` extension for `SQLite`. And you can very easy backup the results in an `.sql` file dump.

Yes thats true, but its not a common way that developers work today. It would be nice if I could just run Docker Compose or something (or use the Testcontainers library), so that when my application starts it also starts a "real" vector database (millvus/chroma/pgvector/etc), populates it with some data, and then runs. When my application stops, the vector database container is torn down. This is a very common development lifecycle and how things work with something like Docker or Testcontainers today. The problem is in this scenario, that the embeddings need to be re-computed each time the application starts up (the same way, for example, Hibernate would re-create the schema(s)/populate data), except computing embeddings can be very time consuming (and expensive ($$)), especially if I'm using a remote embedding model and have to pay per API request.

It would be nice to have some sample pre-computed embeddings in my source tree that can satisfy my local development needs, similar to me having an `import.sql` file in my `src/main/resources` that hibernate knows what to do with, but _only_ when doing local development. This way, when starting up my local development server, I can pass some configuration that tells LangChain4J to bootstrap a vector database container and then load some data into it. That way my application doesn't need to re-compute the embeddings. Otherwise now when my application starts in development mode, it first needs to compute a bunch of embeddings before it stores them in the vector database that was started by Testcontainers.

> 
> You can also not use a Docker container, but start it locally on your computer.
> 
> However, the easiest method is to find an **embedded vector database** (just like SQLite is an embedded database). That would be like a library and store everything in files. But unfortunately, every vector database requires a dedicated server

Using a container and a framework like Testcontainers is pretty common nowadays, so starting a vector database locally on my computer isn't something I should need to do. Many developers don't have permissions on their local computers to install things.

Java Frameworks like Quarkus and Spring, for example, support this development model right out of the box. Thats what I'm trying to solve with this enhancement.

With Quarkus, we are already doing this with the in-memory vector store (see https://docs.quarkiverse.io/quarkus-langchain4j/dev/easy-rag.html#_reusing_embeddings). We'd like to be able to extend the capability to other vector databases through a common api.

As mentioned, this change I'm proposing will not actually implement that common API for all the `EmbeddingStore` interfaces. It will just add some `default` methods to the `EmbeddingStore` interface that throw some `NotYetImplementedException` or something to indicate that a particular implementation doesn't yet support that capability. Then we can implement it later for each implementation. The `InMemoryStore` already implements it.

**Comment by langchain4j** [comment]
Hi @edeandrea, I am not really sure you need to alter `EmbeddingStore` interface in order to implement this feature in Quarkus extension.

All the embeddings can be exported by calling `embeddingStore.search(EmbeddingSearchRequest.builder().queryEmbedding(dummy).maxResults(Integer.MAX_VALUE).build())`

And then imported using `EmbeddingStore.add()` or `addAll()` methods.

**Comment by edeandrea** [comment]
Thank you @langchain4j for the info! I wasn't aware that was there. Let me experiment with it a bit. I'm going to be out on PTO for a couple of weeks, so it may be a bit before I report back.

If you'd prefer to close this issue for now I can reopen it if I have further questions. Or you can leave it open. Up to you!

**Comment by langchain4j** [comment]
@edeandrea ok, no rush! I will close the issue (as we will probably forget about it), but feel free to reopen if needed. Enjoy your PTO! :)

**Comment by LizeRaes** [comment]
@edeandrea sorry for the late reply! I hope the `embeddingstore.search()`method solves your problem, and thanks a lot for your efforts to include 'hot loading' of embedding store changes into Quarkus' awesome dev mode!

**Comment by edeandrea** [comment]
Thanks @LizeRaes ! Maybe I'll have something by the time of our workshop at Devoxx Belgium!
